gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java

```diff
@@ -12,36 +12,81 @@
 
 package gobblin.data.management.retention;
 
+import java.io.Closeable;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
 import java.util.List;
 import java.util.Properties;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
+import com.codahale.metrics.Meter;
+import com.google.common.base.Optional;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
 
 import org.apache.hadoop.fs.FileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import gobblin.configuration.State;
 import gobblin.data.management.retention.dataset.Dataset;
 import gobblin.data.management.retention.dataset.finder.DatasetFinder;
+import gobblin.instrumented.Instrumentable;
+import gobblin.instrumented.Instrumented;
+import gobblin.metrics.GobblinMetrics;
+import gobblin.metrics.MetricContext;
+import gobblin.metrics.Tag;
+import gobblin.util.ExecutorsUtils;
 
 
 /**
  * Finds existing versions of datasets and cleans old or deprecated versions.
  */
-public class DatasetCleaner {
+public class DatasetCleaner implements Instrumentable, Closeable {
 
   public static final String CONFIGURATION_KEY_PREFIX = "gobblin.retention.";
   public static final String DATASET_PROFILE_CLASS_KEY = CONFIGURATION_KEY_PREFIX + "dataset.profile.class";
+  public static final String DATASET_CLEAN_MAX_CONCURRENT_HDFS_CALLS = CONFIGURATION_KEY_PREFIX
+      + "dataset.clean.max.concurrent.hdfs.calls";
+  public static final String DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT = CONFIGURATION_KEY_PREFIX + "hdfs.calls.per.second.limit";
+
+  public static final String DEFAULT_DATASET_CLEAN_MAX_CONCURRENT_HDFS_CALLS = "1000";
+  public static final String DEFAULT_DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT = "10";
```

Comment:
This default seems low. I would rather make it infinite, or at least very large (most users will not change this, and it would suck if the cleaner is slow by default).